test/test_utils/test_reporting.py

```diff
@@ -0,0 +1,290 @@
+        return default
+    else:
+        return markers[0].args[0]
+
+
+class TestReportGenerator:
+    """
+    Class to generate test report files
+    """
+
+    ALLOWED_SINGLE_GPU_TESTS = ("telemetry", "test_framework_version_gpu")
+    SM_REPOS = (
+        "pytorch-training",
+        "pytorch-inference",
+        "tensorflow-tensorflow1_training",
+        "tensorflow-tensorflow2_training",
+        "mxnet-training",
+        "mxnet-inference",
+    )
+    COVERAGE_DOC_EXECUTABLE = "pytest -s --collect-only --generate-coverage-doc"
+
+    def __init__(self, items, test_coverage_file=None, is_sagemaker=False):
+        """
+        The TestReportGenerator requires a list of pytest items. The "generate_coverage_doc" method can be
+        used for most cases
+
+        :param items: list of pytest item objects
+        :param test_coverage_file: optional -- specify the name of the coverage file to write to
+        """
+        self.items = items
+        self.test_coverage_file = get_test_coverage_file_path() if not test_coverage_file else test_coverage_file
+        self.is_sagemaker = is_sagemaker
+        self.failure_conditions = {}
+
+    def handle_single_gpu_instances_test_report(self, function_key, function_keywords, processor="gpu"):
+        """
+        Generally, we do not want tests running on single gpu instances. However, there are exceptions to this rule.
+        This method is used to determine whether we need to raise an error with report generation or not, based on
+        whether we are using single gpu instances or not in a given test function.
+
+        :param function_key: local/path/to/function::function_name
+        :param function_keywords: string of keywords associated with the test function
+        :param processor: whether the test is for cpu, gpu or both
+        :return: processor if not single gpu instance, else "single_gpu", and a dict with updated failure messages
+        """
+
+        # Define conditions where we allow a test function to run with a single gpu instance
+        whitelist_single_gpu = False
+        allowed_single_gpu = self.ALLOWED_SINGLE_GPU_TESTS
+
+        # Regex in order to determine the gpu instance type
+        gpu_instance_pattern = re.compile(r"\w+\.\d*xlarge")
+        gpu_match = gpu_instance_pattern.search(function_keywords)
+
+        if gpu_match:
+            instance_type = gpu_match.group()
+            num_gpus = get_instance_num_gpus(instance_type=instance_type)
+
+            for test in allowed_single_gpu:
+                if test in function_key:
+                    whitelist_single_gpu = True
+                    break
+            if num_gpus == 1:
+                processor = "single_gpu"
+                if not whitelist_single_gpu:
+                    single_gpu_failure_message = (
+                        f"Function uses single-gpu instance type {instance_type}. "
+                        f"Please use multi-gpu instance type."
```

Comment:
It'll also help to add a sentence here that informs that the test can be added to the 'allowed list'.

Reply:
Good catch